xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,223 @@
+      AuthorizationDecision authzDecision = evaluateEngine(allowEngine.conditions.entrySet(), 
+          AuthorizationDecision.Decision.ALLOW, unknownPolicyNames, activation);
+      if (authzDecision != null) {
+        return authzDecision;
+      }
+    }
+    // Only has a denyEngine and it’s unmatched.
+    if (this.allowEngine == null && this.denyEngine != null) {
+      return new AuthorizationDecision(
+          AuthorizationDecision.Decision.ALLOW, new ArrayList<String>());
+    }
+    // None of denyEngine and allowEngine matched, or the single Allow Engine is unmatched.
+    return new AuthorizationDecision(AuthorizationDecision.Decision.DENY, new ArrayList<String>());
+  }
+
+  /** Evaluate a single RbacEngine. */
+  protected AuthorizationDecision evaluateEngine(Set<Map.Entry<String, Expr>> entrySet, 
+      AuthorizationDecision.Decision decision, List<String> unknownPolicyNames, 
+      Activation activation) {
+    for (Map.Entry<String, Expr> condition : entrySet) {
+      try {
+        if (matches(condition.getValue(), activation)) {
+          return new AuthorizationDecision(decision, 
+              new ArrayList<String>(Arrays.asList(new String[] {condition.getKey()})));
+        }
+      } catch (InterpreterException e) {
+        unknownPolicyNames.add(condition.getKey());
+      }
+    }
+    if (unknownPolicyNames.size() > 0) {
+      return new AuthorizationDecision(
+          AuthorizationDecision.Decision.UNKNOWN, unknownPolicyNames);
+    }
+    return null;
+  }
+
+  /** Evaluate if a condition matches the given Enovy Attributes using CEL library. */
+  protected boolean matches(Expr condition, Activation activation) throws InterpreterException {
+    // Set up interpretable used in CEL library's eval function.
+    List<Descriptor> descriptors = new ArrayList<>();
+    RuntimeTypeProvider messageProvider = DescriptorMessageProvider.dynamicMessages(descriptors);
+    Dispatcher dispatcher = DefaultDispatcher.create();
+    Interpreter interpreter = new DefaultInterpreter(messageProvider, dispatcher);
+    Interpretable interpretable = interpreter.createInterpretable(condition);
+    // Parse the generated result object to a boolean variable.
+    try {
+      Object result = interpretable.eval(activation);
+      if (result instanceof Boolean) {
+        return Boolean.valueOf(result.toString());
+      }
+      // Throw an InterpreterException if there are missing Envoy Attributes.
+      if (result instanceof IncompleteData) {
+        throw new InterpreterException.Builder("Incomplete Envoy Attributes to be evaluated.")
+            .build(); 
+      }
+    } catch (InterpreterException e) {
+      // If any InterpreterExceptions are catched, throw it and log the error.
+      log.log(Level.WARNING, e.toString(), e);
+      throw e;
+    }
+    return false;
+  }
+
+  /** Extract Envoy Attributes from EvaluateArgs. */
+  protected ImmutableMap<String, Object> extractFields(EvaluateArgs args) {
```

Comment:
Shouldn't this belong to `EvaluateArgs`? It should know how to generate a Map out of its field values. Otherwise you are exposing fields like `request.method` here. You can call the method (there) `generateEnvoyAttributes` (based on the comment here)

Reply:
You are correct. It makes more sense to move this inside `EvaluateArgs`. Will fix!